tests/63-live-notify_addfd.c

```diff
@@ -0,0 +1,210 @@
+ * published by the Free Software Foundation.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library; if not, see <http://www.gnu.org/licenses>.
+ */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <seccomp.h>
+#include <string.h>
+#include <sys/socket.h>
+#include <sys/wait.h>
+#include <unistd.h>
+
+
+int send_fd(int sock, int fd)
+{
+	struct iovec iov = {.iov_base = "F", .iov_len = 1};
+	char buffer[CMSG_SPACE(sizeof(fd))]; // Do i need to set it to zero?
+	memset(buffer, 0, sizeof(buffer));
+
+	struct msghdr msg = {
+		.msg_iov = &iov,
+		.msg_iovlen = 1,
+		.msg_control = buffer,
+		.msg_controllen = sizeof(buffer)
+	};
+
+	struct cmsghdr *cmsg = CMSG_FIRSTHDR(&msg);
+	cmsg->cmsg_level = SOL_SOCKET;
+	cmsg->cmsg_type = SCM_RIGHTS;
+	cmsg->cmsg_len = CMSG_LEN(sizeof(fd));
+
+	memcpy(CMSG_DATA(cmsg), &fd, sizeof(fd));
+
+	return sendmsg(sock, &msg, 0);
+}
+
+int recv_fd(int sock)
+{
+	char m_buffer[1];
+	struct iovec iov = {.iov_base = m_buffer, .iov_len = 1};
+
+	char buffer[CMSG_SPACE(sizeof(int))];
+	struct msghdr msg = {
+		.msg_iov = &iov,
+		.msg_iovlen = 1,
+		.msg_control = buffer,
+		.msg_controllen = sizeof(buffer)
+	};
+	struct cmsghdr *cmsg = CMSG_FIRSTHDR(&msg);
+
+	if (recvmsg(sock, &msg, 0) < 0)
+		return -1;
+
+	int fd;
+	memcpy(&fd, CMSG_DATA(cmsg), sizeof(fd));
+	return fd;
+}
+
+int main(int argc, char *argv[])
+{
+	int rc, status;
+	int sock_pair[2];
+	int notify_fd = -1, new_fd = -1;
+	struct seccomp_notif *req = NULL;
+	struct seccomp_notif_resp *resp = NULL;
+	struct seccomp_notif_addfd addfd = {0};
+	scmp_filter_ctx ctx = NULL;
+	pid_t pid = 0;
+
+	ctx = seccomp_init(SCMP_ACT_ALLOW);
+	if (ctx == NULL)
+		return ENOMEM;
+
+	rc = seccomp_rule_add(ctx, SCMP_ACT_NOTIFY, SCMP_SYS(openat), 0, NULL);
+	if (rc)
+		goto out;
+
+	// set up socket pair for sending notify_fd
+	rc = socketpair(AF_UNIX, SOCK_SEQPACKET, 0, sock_pair);
+	if (rc < 0) {
+		rc = -errno;
+		goto out;
+	}
+
+	pid = fork();
+	if (pid == 0) {
+		close(sock_pair[0]); // close the parent's end
+
+		rc = seccomp_load(ctx);
+		if (rc < 0)
+			goto out;
+
+		rc = seccomp_notify_fd(ctx);
+		if (rc < 0)
+			goto out;
+		notify_fd = rc;
+
+		rc = send_fd(sock_pair[1], notify_fd);
+		if (rc < 0) {
+			rc = -errno;
+			goto out;
+		}
+		close(notify_fd);
+
+		int ret = openat(AT_FDCWD, "/etc/hostname", O_RDONLY);
+		if (ret < 0) {
+			exit(ret);
+		}
+
+		char buf[128];
+		ssize_t bytes_read = read(ret, buf, sizeof(buf));
+
+		close(ret);
+		close(sock_pair[1]);
+		exit(bytes_read); // bytes_read should be 0, as it's reading /dev/null
+	} else {
+		close(sock_pair[1]); // close the child's end
```

Comment:
I would recommend moving the parent code to its own function.  This will improve readability and better scope the variables that are only needed by the parent process.

And, as outlined above, it will fix the erroneous `goto` logic in the child, as you'll have to handle that properly in its own function